fix(Core/Grids): Grid improvements

Takenbacon opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

**!! PR is WIP and not ready to be merged**

1. Dynamically allocates grid cells. At (64 * 64) * (8 * 8) = **262,144 cells per map** improving the memory efficiency of the structure has a significant impact considering most cells are usually empty anyways. When compiled in RelWithDebug and using the `PreloadAllNonInstancedMapGrids = 1` config option, RAM usage decreased from 9,810MB to 4,645MB (no mmaps) and startup time improved from 19 seconds to 16 seconds.
2. Greatly simplify ObjectGridLoader



While the changes as of this moment are testable, there is more I'd still like to do, thus PR is WIP for now.

## Changes Proposed:
<!-- If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address. -->
This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes 

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->
The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
This PR has been:
- [ ] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.


## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
4.
5.

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- [ ]
- [ ]

<!-- If you intend to contribute repeatedly to our project, it is a good idea to join our discord channel. We set ranks for our contributors and give them access to special resources or knowledge: https://discord.com/invite/GyFvXpk7)
     Do not remove the instructions below about testing, they will help users to test your PR -->
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

Grimdhex commented:
One small change if you want for more readability: go out the GridMap inside the Map file. And put it in this own file.

Takenbacon commented:
> One small change if you want for more readability: go out the GridMap inside the Map file. And put it in this own file.

Yeah I saw that, I'm currently weighing how much I should change in a single PR, and also considering this is specifically marked "grids" whereas some of my changes would be more related to map class.

Kitzunu commented:
Interesting 👍

Grimdhex commented on src/server/game/Maps/Map.cpp:
```diff
@@ -447,7 +443,7 @@ void Map::EnsureGridCreated_i(const GridCoord& p)
     {
         // pussywizard: moved setNGrid to the end of the function
-        NGridType* ngt = new NGridType(p.x_coord * MAX_NUMBER_OF_GRIDS + p.y_coord, p.x_coord, p.y_coord);
+        NGridType* ngt = new NGridType(p.y_coord * MAX_NUMBER_OF_GRIDS + p.x_coord, p.x_coord, p.y_coord);
```
Why reverse the coordinates here?

Takenbacon replied:
To synchronize NGrid ID with CoordPair's ID which enables the ObjectGridLoader simplifications. I believe this to have just been a simple mistake in the original grid system implementation as GridId isn't used much let alone in conjunction with a grid CoordPair. Previously, spawned objects were stored by cell ID and the object grid loader would loop all cells in the grid and do a lookup on the _mapObjectGuidsStore map for each cell in the grid. Since we're always loading every cell in a grid anyways, it's easy to cut out the bloat and just store the objects by grid ID instead, simplifying the process.

CoordPair:
`[[nodiscard]] uint32 GetId() const
{ return y_coord * LIMIT + x_coord; }`

Grimdhex commented:
I see that you have moved some grid definitions outside of maps. If you don't these data are duplicated at differents places and can be merged. I've a PR that port these changes available only on TC master branch (https://github.com/TrinityCore/TrinityCore/commit/65f82c7c274a7b7058c065dfffe224c59855f41d#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52)

Takenbacon commented:
> I see that you have moved some grid definitions outside of maps. If you don't these data are duplicated at different places in the framework, and it can be merged. I've a PR that port these changes available only on TC master branch ([TrinityCore/TrinityCore@65f82c7#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52](https://github.com/TrinityCore/TrinityCore/commit/65f82c7c274a7b7058c065dfffe224c59855f41d#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52)); It's just for information, not necessary for this PR.

Sorry, I don't quite understand. Are you saying you want me to merge in that TC commit?

Grimdhex commented:
> > I see that you have moved some grid definitions outside of maps. If you don't these data are duplicated at different places in the framework, and it can be merged. I've a PR that port these changes available only on TC master branch ([TrinityCore/TrinityCore@65f82c7#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52](https://github.com/TrinityCore/TrinityCore/commit/65f82c7c274a7b7058c065dfffe224c59855f41d#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52)); It's just for information, not necessary for this PR.
> 
> 
> 
> Sorry, I don't quite understand. Are you saying you want me to merge in that TC commit?

Not necessary. If you want other improvements for grids, it's another possibility. But no real impact IG. Just code management.

Takenbacon commented:
> > > I see that you have moved some grid definitions outside of maps. If you don't these data are duplicated at different places in the framework, and it can be merged. I've a PR that port these changes available only on TC master branch ([TrinityCore/TrinityCore@65f82c7#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52](https://github.com/TrinityCore/TrinityCore/commit/65f82c7c274a7b7058c065dfffe224c59855f41d#diff-ad66842f3a1bb1684807e622c9f129d330fa7cd3daf4ba9a62254cd2de340c52)); It's just for information, not necessary for this PR.
> > 
> > 
> > Sorry, I don't quite understand. Are you saying you want me to merge in that TC commit?
> 
> Not necessary. If you want other improvements for grids, it's another possibility. But no real impact IG. Just code management.

Ah gotcha. Yeah, I still have a bunch of local WIP changes to reorganize some more of the code. But I am definitely starting to get tired of messing with it so I might just end up finishing what I'm doing and calling it "good" for now, haha.

Nyeriah commented:
Every small improvement is a good improvement, absolutelly no need to address all that in one go 😅 Thanks for this all other PRs btw.

Takenbacon commented:
My most recent commit will be the bulk of the refactoring I'll do in regards to grids I think. There's still a couple smaller things I'm planning on looking into cleaning up/improving but I could probably spend the rest of my life on this and I'd like to move on to other things eventually haha. It does compile currently (on MSVC) and I was able to login and test a couple things. I will be out of town this upcoming weekend so I'll hopefully get this wrapped up the following if all goes well.

Takenbacon commented:
I FINALLY got the build fixed on gcc/clang. Holy crap. I really just need to get unlazy and set up my own environment instead of depending on the automated build scripts lol. Anyways, that should be it for this PR in terms of changes. Ready to be tested and WIP tag removed when convenient.

Takenbacon commented on src/server/game/Grids/GridTerrainData.h:
```diff
@@ -0,0 +1,193 @@
+class GridTerrainData
+{
+    uint32  _flags;
+    union
```
Not sure if variant works too well in this particular case with pointers to arrays. Either way, I've done some refactoring to the file and removed the union into separate structures, as it makes a little more sense as to how the data is loaded anyways.

Takenbacon commented:
Made a couple small changes and a fix.

Interesting tidbit: Previously (before my PR), when a grid terrain map failed to load for any reason the GridMap remained created and accessible - either in a partially loaded or default-initialized state, which I think is not great and could've led to some odd behavior. I fixed that but it did have an unintended consequence of at least in one case in `Map::GetWaterOrGroundLevel` having a different default return value causing an issue with random movement in areas without terrain map data (RFC instance for example). I've gone through all the uses of `GetGridTerrainData` and think I got all the uses checked now, I hope.

I've done about as much local testing as I can, just looking for others to now give it a test and feedback results so this PR hopefully doesn't stay in an idle state for too long of a duration.

Kitzunu commented:
@azerothcore/testers Do your thing

blinkysc commented:
Been trying this out for a few hours, seems like a very big improvement! Haven't encountered any bugs. Also no joke with the memory improvements...

before
![before](https://github.com/user-attachments/assets/fcaa0201-d870-4efc-92d8-3343b9ff807f)

after
![after](https://github.com/user-attachments/assets/ee0a80a8-69f4-490c-9f2a-b3016f0fe91e)

Also some perf reports (I was interested if many more cycles or anything else) 
[before_perf.txt](https://github.com/user-attachments/files/18776028/before_perf.txt)
[after_perf.txt](https://github.com/user-attachments/files/18776029/after_perf.txt)

I wanted to try with many thousand bots but seems many of the cpp/h changed runs into changes to the playerbot fork. May take awhile to try and "plug and play" those difference to give it a go with that.

PkllonG commented:
![image](https://github.com/user-attachments/assets/72f7de50-5edb-433a-bb26-b86384808d97)
Before PR, they are lit

Kitzunu commented:
Conflict

Takenbacon commented:
> ![image](https://private-user-images.githubusercontent.com/18535853/412744733-72f7de50-5edb-433a-bb26-b86384808d97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NjI1OTUsIm5iZiI6MTczOTQ2MjI5NSwicGF0aCI6Ii8xODUzNTg1My80MTI3NDQ3MzMtNzJmN2RlNTAtNWVkYi00MzNhLWJiMjYtYjg2Mzg0ODA4ZDk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE1NTgxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI4MjQzMmMzZDc2YTdkM2U3Yzc0MzcwZjk1NzBkNmJlYTczYjAxNTk0NWM2YmU5YjEyY2VlM2Q5OWY5YTg3M2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xIWHnAgq-enYr54zAxYZJ82r113WQepGJvQ1Gn1xUZI) Before PR, they should be lit
> 
> Before PR ![image](https://private-user-images.githubusercontent.com/18535853/412750853-c2487818-7bdd-496d-b461-18f570f2b994.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NjI1OTUsIm5iZiI6MTczOTQ2MjI5NSwicGF0aCI6Ii8xODUzNTg1My80MTI3NTA4NTMtYzI0ODc4MTgtN2JkZC00OTZkLWI0NjEtMThmNTcwZjJiOTk0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE1NTgxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5YTNkYWU1ZjhlOGMxOTc4Mzk3OTk1NDY1N2ZiYTZhNzQ1OWM5NmExNzI4NmU0OTk5ZmE1ZWEyYjk5YzExYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GWFMkuk9S74hsti7O97uUIJywEMNYUmhED6b7Tr4cAg)

Fixed, thank you for checking. I was loading grids a little too early in the process before instance scripts so they weren't being properly assigned to the script. Good catch.

> Conflict

Fixed

PkllonG commented:
I ran it for a while longer and didn't find any errors for a while

PkllonG commented:
So far, it's been working well

Nyeriah commented:
I will need some help with this one. We are populating maps before they have any players inside now, so instance maps can no longer tell wether they should spawn one NPC or another, where the spawn is conditioned by faction inside the instance

https://github.com/azerothcore/azerothcore-wotlk/issues/21529

Takenbacon commented:
> I will need some help with this one. We are populating maps before they have any players inside now, so instance maps can no longer tell wether they should spawn one NPC or another, where the spawn is conditioned by faction inside the instance
> 
> #21529

Ah I had forgotten about faction instances, I'm guessing the hook that is setting faction Id is being called after the grids get loaded. Should be a fairly straightforward fix, either moving the hook call or maybe alternatively moving it to a different hook (OnInstanceCreate?) which might be a better solution anyways. I will take a look in the morning after work if possible.

Nyeriah commented:
I tried looking into it, but the issue is that the instance is always empty when they are spawned, so we would need to add players before the creatures are spawned (?). Or get that information during the creation process itself, when the map is created for the first player to enter

Takenbacon commented:
> I tried looking into it, but the issue is that the instance is always empty when they are spawned, so we would need to add players before the creatures are spawned (?). Or get that information during the creation process itself, when the map is created for the first player to enter (but we don't have the instance script initialized at that point?)

https://github.com/azerothcore/azerothcore-wotlk/blob/28f5ead7f308a6a301d3dd8798c5031e8a52dcb1/src/server/game/Maps/MapInstanced.cpp#L216

Grids are spawned on the hook here, right after instance script is created. Probably need to pass player to this function and add a new hook just before grids are loaded.

Takenbacon commented:
@Nyeriah https://github.com/Takenbacon/azerothcore-wotlk/commit/e607d1cde037214a4229577efe01e3a69ef33e2a I threw this together really quickly if you or someone want to test, it should work or at the very least lead someone in the right direction, I just unfortunately don't have enough time until after my work week is done, if it's still a problem I can look at it this upcoming weekend but I am quite confident it should take care of it.
